engine/crates/runtime-local/src/rate_limiting/in_memory/key_based.rs

```diff
@@ -7,69 +6,48 @@ use futures_util::future::BoxFuture;
 use futures_util::FutureExt;
 use governor::Quota;
 use grafbase_telemetry::span::GRAFBASE_TARGET;
-use serde_json::Value;
 
-use http::{HeaderName, HeaderValue};
-use runtime::rate_limiting::{Error, GraphRateLimit, KeyedRateLimitConfig, RateLimiter, RateLimiterContext};
-use tokio::sync::mpsc;
+use runtime::rate_limiting::{Error, GraphRateLimit, RateLimitKey, RateLimiter, RateLimiterContext};
+use tokio::sync::watch;
 
-pub struct RateLimitingContext(pub String);
-
-impl RateLimiterContext for RateLimitingContext {
-    fn header(&self, _name: HeaderName) -> Option<&HeaderValue> {
-        None
-    }
-
-    fn graphql_operation_name(&self) -> Option<&str> {
-        None
-    }
-
-    fn ip(&self) -> Option<IpAddr> {
-        None
-    }
-
-    fn jwt_claim(&self, _key: &str) -> Option<&Value> {
-        None
-    }
-
-    fn key(&self) -> Option<&str> {
-        Some(&self.0)
-    }
-}
+type Limits = HashMap<RateLimitKey<'static>, GraphRateLimit>;
+type Limiters = HashMap<RateLimitKey<'static>, governor::DefaultKeyedRateLimiter<usize>>;
 
 pub struct InMemoryRateLimiter {
-    limiters: Arc<RwLock<HashMap<String, governor::DefaultKeyedRateLimiter<usize>>>>,
+    limiters: Arc<RwLock<Limiters>>,
 }
 
 impl InMemoryRateLimiter {
-    pub fn runtime(
-        config: KeyedRateLimitConfig,
-        mut updates: mpsc::Receiver<HashMap<String, GraphRateLimit>>,
-    ) -> RateLimiter {
+    pub fn runtime(mut updates: watch::Receiver<Limits>) -> RateLimiter {
         let mut limiters = HashMap::new();
 
         // add subgraph rate limiting configuration
-        for (name, config) in config.rate_limiting_configs {
-            let Some(limiter) = create_limiter(config) else {
+        for (name, config) in updates.borrow_and_update().iter() {
+            let Some(limiter) = create_limiter(*config) else {
```

Comment:
:zap: **Suggestion** 
Ensure that the `create_limiter` function handles the case where `config` might be invalid or not as expected, to avoid potential runtime errors.